libbeat/publisher/pipeline/acker.go

```diff
@@ -400,37 +404,57 @@ func (a *eventDataACK) onACK(total, acked int) {
 type waitACK struct {
 	acker acker
 
-	signal    chan struct{}
-	waitClose time.Duration
+	signalAll  chan struct{} // ack loop notifies `close` that all events have been acked
+	signalDone chan struct{} // shutdown handler telling `wait` that shutdown has been completed
+	waitClose  time.Duration
 
 	active atomic.Bool
 
 	// number of active events
 	events atomic.Uint64
+
+	afterClose func()
 }
 
-func newWaitACK(acker acker, timeout time.Duration) *waitACK {
+func newWaitACK(acker acker, timeout time.Duration, afterClose func()) *waitACK {
 	return &waitACK{
-		acker:     acker,
-		signal:    make(chan struct{}, 1),
-		waitClose: timeout,
-		active:    atomic.MakeBool(true),
+		acker:      acker,
+		signalAll:  make(chan struct{}, 1),
+		signalDone: make(chan struct{}),
```

Comment:
I am gonna start using `signal` prefix in my code, I see how you are using it and I think it made code clearer.